src/perf/lib/PerfHelpers.h

```diff
@@ -0,0 +1,629 @@
+    }
+    QuicAddr(const QuicAddr &Addr, uint16_t Port) {
+        SockAddr = Addr.SockAddr;
+        QuicAddrSetPort(&SockAddr, Port);
+    }
+    uint16_t GetPort() const { return QuicAddrGetPort(&SockAddr); }
+};
+
+template<typename T>
+class UniquePtr {
+public:
+    UniquePtr() noexcept = default;
+
+    explicit UniquePtr(T* _ptr) :
+        ptr{_ptr}
+    {
+    }
+
+    UniquePtr(const UniquePtr& other) = delete;
+    UniquePtr& operator=(const UniquePtr& other) = delete;
+
+    UniquePtr(UniquePtr&& other) noexcept {
+        this->ptr = other->ptr;
+        other->ptr = nullptr;
+    }
+
+    UniquePtr& operator=(UniquePtr&& other) noexcept {
+        if (this->ptr) {
+            delete this->ptr;
+        }
+        this->ptr = other->ptr;
+        other->ptr = nullptr;
+    }
+
+    ~UniquePtr() noexcept {
+        if (this->ptr) {
+            delete this->ptr;
+        }
+    }
+
+    void reset(T* lptr) noexcept {
+        if (this->ptr) {
+            delete this->ptr;
+        }
+        this->ptr = lptr;
+    }
+
+    T* release() noexcept {
+        T* tmp = ptr;
+        ptr = nullptr;
+        return tmp;
+    }
+
+    T* get() const noexcept { return ptr; }
+
+    T& operator*() const { return *ptr; }
+    T* operator->() const noexcept { return ptr; }
+    operator bool() const noexcept { return ptr != nullptr; }
+    bool operator == (T* _ptr) const noexcept { return ptr == _ptr; }
+    bool operator != (T* _ptr) const noexcept { return ptr != _ptr; }
+
+private:
+    T* ptr = nullptr;
+};
+
+template<typename T>
+class UniquePtr<T[]> {
+public:
+    UniquePtr() noexcept = default;
+
+    explicit UniquePtr(T* _ptr) :
+        ptr{_ptr}
+    {
+    }
+
+    UniquePtr(const UniquePtr& other) = delete;
+    UniquePtr& operator=(const UniquePtr& other) = delete;
+
+    UniquePtr(UniquePtr&& other) noexcept {
+        this->ptr = other->ptr;
+        other->ptr = nullptr;
+    }
+
+    UniquePtr& operator=(UniquePtr&& other) noexcept {
+        if (this->ptr) {
+            delete[] this->ptr;
+        }
+        this->ptr = other->ptr;
+        other->ptr = nullptr;
+    }
+
+    ~UniquePtr() noexcept {
+        if (this->ptr) {
+            delete[] this->ptr;
+        }
+    }
+
+    void reset(T* _ptr) noexcept {
+        if (this->ptr) {
+            delete[] this->ptr;
+        }
+        this->ptr = _ptr;
+    }
+
+    T* release() noexcept {
+        T* tmp = ptr;
+        ptr = nullptr;
+        return tmp;
+    }
+
+    T* get() const noexcept { return ptr; }
+
+    T& operator[](size_t i) const {
+        return *(ptr + i);
+    }
+
+    operator bool() const noexcept { return ptr != nullptr; }
+    bool operator == (T* _ptr) const noexcept { return ptr == _ptr; }
+    bool operator != (T* _ptr) const noexcept { return ptr != _ptr; }
+
+private:
+    T* ptr = nullptr;
+};
+
+inline
+int
+WriteOutput(
+    _In_z_ const char* format
+    ...
+    )
+{
+#ifndef _KERNEL_MODE
```

Comment:
Just always write to the buffer independent of platform.

Reply:
I still don't have a buffer yet. I want to do it right, so I'll do that when I do kernel mode. But when I do I'll use the buffer for both.